File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
  const heap = []
  for (const n of nums) {
    let [l, r] = [0, heap.length]
    while (l < r) {
      const m = ~~((l + r) / 2)
      if (heap[m] < n) l = m + 1
      else r = m
    }
    if (l >= heap.length) heap.push(n)
    else heap[l] = n
  }
  return heap.length
};

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.go
func lengthOfLIS(nums []int) int {
  l := len(nums)
  heap := make([]int, l)
  heap[0] = nums[0]
  result := 1
  for i := 1; i < l; i++ {
    target := nums[i]
    l, r := 0, result
    for l < r {
      m := (l + r) / 2
      if heap[m] < target {
        l = m + 1
      } else {
        r = m
      }
    }  
    if l >= result {
      result++
    }
    heap[l] = target
  }
  return result
}